includes/utils/class-amp-validation-utils.php

```diff
@@ -303,6 +303,97 @@ public static function add_validation_hooks() {
 
 		add_filter( 'do_shortcode_tag', array( __CLASS__, 'decorate_shortcode_source' ), -1, 2 );
 		add_filter( 'amp_content_sanitizers', array( __CLASS__, 'add_validation_callback' ) );
+
+		$is_gutenberg_active = (
+			function_exists( 'do_blocks' )
+			&&
+			function_exists( 'gutenberg_parse_blocks' )
+			&&
+			function_exists( 'gutenberg_render_block' )
+			&&
+			class_exists( 'WP_Block_Type_Registry' )
+		);
+		if ( $is_gutenberg_active ) {
+			$do_blocks_priority = has_filter( 'the_content', 'do_blocks' );
+			if ( false !== $do_blocks_priority ) {
+				remove_filter( 'the_content', 'do_blocks', $do_blocks_priority );
+				add_filter( 'the_content', array( __CLASS__, 'do_blocks' ), $do_blocks_priority );
+			}
+		}
+	}
+
+	/**
+	 * Parse blocks out of content and render with AMP source stack comments added to demarcate blocks.
+	 *
+	 * @see do_blocks() prior to https://github.com/WordPress/gutenberg/commit/49f05e53a4e9ceff0c08fe646f3b0bb724f58b53
+	 *
+	 * @param string $content Content.
+	 * @return string Content with rendered blocks.
+	 */
+	public static function do_blocks( $content ) {
+		$blocks = gutenberg_parse_blocks( $content );
+		return self::render_blocks( $blocks );
+	}
+
+	/**
+	 * Render parsed blocks.
+	 *
+	 * @param array $blocks Blocks.
+	 * @return string Rendered blocks.
+	 */
+	protected static function render_blocks( $blocks ) {
+		$rendered_blocks = '';
+
+		foreach ( $blocks as $block ) {
+			$rendered_block = gutenberg_render_block( $block );
+
+			// Obtain source information for block.
+			$source = null;
+			if ( ! empty( $block['blockName'] ) ) {
+				$source = array(
+					'block_name' => $block['blockName'],
+				);
+				if ( ! empty( $block['attrs'] ) ) {
+					$source['block_attrs'] = $block['attrs'];
+				}
+				$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+				if ( $block_type && $block_type->is_dynamic() ) {
+					$callback_source = self::get_source( $block_type->render_callback );
+					if ( $callback_source ) {
+						$source = array_merge(
+							$source,
+							$callback_source
+						);
+					}
+				}
+			}
+
+			// Prepend rendered block with start source comment.
+			if ( $source ) {
+				$rendered_block = self::get_source_comment( $source, true ) . $rendered_block;
+			}
+
+			// Render nested blocks.
+			if ( ! empty( $block['innerBlocks'] ) ) {
+				$rendered_inner_blocks = self::render_blocks( $block['innerBlocks'] );
+
+				// Inject rendered inner blocks before closing tag of outer block (or at end of string if no closing). @todo Confirm approach here.
```

Comment:
I've raised this in WordPress `#core-editor`:

![image](https://user-images.githubusercontent.com/134745/37415027-b7a4ac1a-2767-11e8-8892-f7da8aebd8d3.png)


https://wordpress.slack.com/archives/C02QB2JS7/p1521043684000458